File: Problems/0018_4Sum/Project_Python3/4Sum.py
# coding: utf-8

import os
import sys
import time
import numpy as np

class Solution:
    def fourSum(self, nums, target):
        nums.sort()
        results = []
        self.findNsum(nums, target, 4, [], results)
        return results

    def findNsum(self, nums, target, N, result, results):
        if len(nums) < N or N < 2: return

        # solve 2-sum
        if N == 2:
            l, r = 0, len(nums)-1
            while l < r:
                if nums[l] + nums[r] == target:
                    results.append(result + [nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                    while r > l and nums[r] == nums[r + 1]:
                        r -= 1
                elif nums[l] + nums[r] < target:
                    l += 1
                else:
                    r -= 1
        else:
            for i in range(0, len(nums) - N + 1):   # careful about range
                if target < nums[i]*N or target > nums[-1]*N:  # take advantages of sorted list
                    break
                if i == 0 or i > 0 and nums[i - 1] != nums[i]:  # recursively reduce N
                    self.findNsum(nums[i + 1:], target-nums[i], N - 1, result+[nums[i]], results)
        return

    def fourSum_work(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        results = []
        nums.sort()
        for i in range(len(nums) - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            if target >= 0:
                if nums[i] > target:
                    break
            for j in range(i + 1, len(nums) - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                for k in range(j + 1, len(nums) - 1):
                    if k > j + 1 and nums[k] == nums[k - 1]:
                        continue
                    for l in range(k + 1, len(nums)):
                        if l > k + 1 and nums[l] == nums[l - 1]:
                            continue
                        if nums[i] + nums[j] + nums[k] + nums[l] == target:
                            temp_arr = [nums[i], nums[j], nums[k], nums[l]]
                            temp_arr.sort()
                            results.append(temp_arr)
        return results

def str_to_int_array(flds):
    nums = [0]*len(flds)
    for i in range(len(flds)):
        nums[i] = int(flds[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace(" ","").replace("[[","").replace("]]","").rstrip().split("],[")
    nums = str_to_int_array(flds[0].split(","))
    target = int(flds[1])
    print("nums = %s" %nums)
    print("target = %s" %target)

    time0 = time.time()

    sl = Solution()
    result = sl.fourSum(nums, target)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()

File: Problems/0001_0099/0018_4Sum/Project_CS/4Sum.cs
using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

public class Solution
{
    public List<List<int>> FourSum(int[] nums, int target)
    {
        Array.Sort(nums);
        List<List<int>> results = new List<List<int>>();
        List<int> result = new List<int>();

        findNsum(nums, target, 4, result, results);
        return results;
    }

    public void findNsum(int[] nums, int target, int N, List<int> result, List<List<int>> results)
    {
        if (nums.Length < N || N < 2)
            return;

        // solve 2-sum
        if (N == 2)
        {
            int l = 0;
            int r = nums.Length - 1;
            while (l < r)
            {
                if (nums[l] + nums[r] == target)
                {
                    result.Add(nums[l]);
                    result.Add(nums[r]);
                    results.Add(result);
                    l++;
                    r--;
                    while (l < r && nums[l] == nums[l - 1])
                        l++;
                    while (r > l && nums[r] == nums[r + 1])
                        r--;
                }
                else if (nums[l] + nums[r] < target)
                    l++;
                else
                    r--;
            }
        }
        else
        {
            for (int i = 0;  i < nums.Length - N + 1; ++i)    // careful about range
            {
                if (target < nums[i]*N || target > nums[nums.Length - 1]*N)  //take advantages of sorted list
                    break;
                if (i == 0 || i > 0 && nums[i - 1] != nums[i])  // recursively reduce N
                {
                    int[] temp = new int[nums.Length - (i + 1)];
                    Array.Copy(nums, i + 1, temp, 0, nums.Length - (i + 1));

                    List<int> result_temp = new List<int>(result);
                    result_temp.Add(nums[i]);

                    findNsum(temp, target-nums[i], N - 1, result_temp, results);
                }
            }
        }

        return;
    }

    public int[] str_to_int_array(string s)
    {
        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0].ToString();
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i].ToString();
        }

        return results;
    }

    public string output_IList_array(List<List<int>> flds)
    {
        string results = "";
        if (flds.Count <= 0)
            return results;
        
        results = "[";
        for (int i = 0; i < flds.Count; ++i)
        {
            results += "[" + flds[i][0].ToString();
            for (int n = 1; n < flds[i].Count; ++n)
                results += ", " + flds[i][n].ToString();
            if (i < flds.Count - 1)
                results += "],";
            else
                results += "]";
        }

        return results + "]";
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[[","").Replace("]]","").Trim().Split("],[", StringSplitOptions.None);
        int[] nums = str_to_int_array(flds[0]);
        int target = int.Parse(flds[1]);
        Console.WriteLine("nums = " + output_int_array(nums));
        Console.WriteLine("target = " + target);

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        List<List<int>> results = FourSum(nums, target);
        sw.Stop();

        Console.WriteLine("result = " + output_IList_array(results));
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
